### ISSUE TITLE
WeaveModel becomes a memory hog, gives seg fault

### ISSUE THREAD

**Issue by 1particle** [deepchem/deepchem#1279]
I have been able, in the past, to run WeaveModel on a small dataset of peptides with their MIC values as the target. Trying to repeat that partial success on a dataset of larger proteins (most of them around 70 aa in length) has completely failed, and I get segmentation faults all the time.

I run the CPU version of deepchem (version 2.0) on the [Niagara](https://docs.computecanada.ca/wiki/Niagara) cluster of Compute Canada, with over **200 GB** of memory per job. My code is a minimally altered version of the [tox21 example](https://github.com/deepchem/deepchem/blob/master/examples/tox21/tox21_tensorgraph_weave.py), and my model specification line is

`model = WeaveTensorGraph(len(tasks), n_atom_feat=75,n_pair_feat=14, n_hidden=50, learning_rate=0.0001, n_graph_feat=256, dropout=0.500, momentum=0.800, mode="regression")
`

I have used batch sizes as small as 10 and as large as 100, and the result is always the same: featurization completes and the training fails at the first epoch, with a segmentation faults and a massive core dump. Having tried to rule out all other causes, I can only assume that the WeaveModel is prohibitively expensive for proteins of even moderate size.

My main question is: can one avoid the segmentation faults when using WeaveModel on larger proteins, even if at the expense of slower trainings? In my experiment with antimicrobial peptides, I found WeaveModel to be superior to graphconv models of various depths and widths, so I hope I can also use the WeaveModel on larger proteins.

**Comment by rbharath** [comment]
Hmm, interesting. I think the segfault might be caused by the input data, since the Tox21 example looks to be running fine.

Would it be possible to share a failing datapoint? I think the segfault might be coming from RDKit, but hard to say for sure.

CC @miaecle

**Comment by 1particle** [comment]
Here's a sample dataset that causes the WeaveModel to halt by a segmentation fault as I explained above:
[Synthetic_large_AMPs__SMILES.zip](https://github.com/deepchem/deepchem/files/2069649/Synthetic_large_AMPs__SMILES.zip)
The data is synthetic (i.e., it's generated from duplicating/triplicating peptide sequences) and the target values are assigned randomly, but it's valid data for training as far as I can say.

**Comment by rbharath** [comment]
@1particle Would you be able to put together a small test script that uses this dataset and reproduces the seg-fault? We can use that to start debugging on our end.

**Comment by peastman** [comment]
I've been looking into this.  I can reproduce the problem, but it's not clear to me if this is a bug or just inherent in how Weave works.  When I run the example, several of the input arrays for the first batch are truly enormous.  Their first dimension is size 18414489.  The dataset is very small, but `WeaveModel.default_generator()` generates arrays for each molecule whose size is proportional to `n_atoms**2`.  Since the molecules are hundreds of atoms in size, that adds up quickly.

Could the model be implemented in a different way that would use less memory?  Or is this inherent in how it works?

**Comment by 1particle** [comment]
Whether we call this a bug or a feature, it's an issue that renders the weave model (as currently implemented) inapplicable to larger molecules. And this bothers me because the weave model (in my experience) performs better on small and rather noisy datasets of large molecules than the graph conv model.
A new implementation, with this issue in mind, is probably the only way to solve this, if somebody is kind enough!

**Comment by peastman** [comment]
That's my question.  Is it possible for a new implementation to fix it?  Or is it just inherent in how weave works that its memory has to scale as O(n^2) in the number of atoms?

**Comment by mmahmedgithub** [comment]
Hi Mostafa, just wondering if you can send me the script file for such big jobs you are running. I am having similar problems with compute canada on Graham. I am using the conda version of deepchem for my jobs. Thanks

**Comment by 1particle** [comment]
It's posted above.

**Comment by mmahmedgithub** [comment]
this is the python file. I mean the batch script file, otherwise you are running the jobs interactively on the head node.

**Comment by rbharath** [comment]
Closing this old issue. We've made several improvements to WeaveModel and it runs comfortably on larger molecules now
